app/src/main/java/org/p2p/wallet/send/interactor/usecase/CalculateSendFeesUseCase.kt

```diff
@@ -0,0 +1,194 @@
+import org.p2p.wallet.feerelayer.model.FeeRelayerFee
+import org.p2p.wallet.rpc.interactor.TransactionAddressInteractor
+import org.p2p.wallet.rpc.repository.amount.RpcAmountRepository
+import org.p2p.wallet.send.model.SearchResult
+import org.p2p.wallet.send.repository.SendServiceRepository
+import org.p2p.wallet.utils.toPublicKey
+
+class CalculateSendFeesUseCase(
+    private val amountRepository: RpcAmountRepository,
+    private val sendServiceRepository: SendServiceRepository,
+    private val getFeesInPayingTokenUseCase: GetFeesInPayingTokenUseCase,
+    private val addressInteractor: TransactionAddressInteractor,
+) {
+
+    suspend fun execute(
+        sourceToken: Token.Active,
+        feePayerToken: Token.Active,
+        searchResult: SearchResult,
+        @Suppress("UNUSED_PARAMETER") useCache: Boolean = true
+    ): FeeCalculationState {
+        return try {
+            val minRentExemption = getMinRentExemption(sourceToken)
+            val alternativeMinRentExemption = getSendServiceMinRentExemptionForSameTokens(sourceToken, feePayerToken)
+
+            val feesInSol = getTransactionFeesInSol(
+                minRentExemption = minRentExemption,
+                sourceToken = sourceToken,
+                feePayerToken = feePayerToken,
+                searchResult = searchResult
+            )
+
+            if (feesInSol.totalFeeLamports.isZero()) {
+                Timber.i("Total fees are zero!")
+                return FeeCalculationState.NoFees
+            }
+
+            val feesInFeePayerToken = getTransactionFeesInToken(
+                targetToken = feePayerToken,
+                feesInSol = feesInSol,
+                alternativeAccountCreationFee = alternativeMinRentExemption
+            )
+
+            val feesInSourceToken = getTransactionFeesInToken(
+                targetToken = sourceToken,
+                feesInSol = feesInSol,
+                alternativeAccountCreationFee = alternativeMinRentExemption
+            )
+
+            // it is incorrect to return fees in sol if there is some error happened
+            // because we would add apples to oranges when choosing fee payer token
+            require(feesInFeePayerToken != null && feesInSourceToken != null) {
+                buildString {
+                    append("Cannot calculate transaction fees in source (${sourceToken.tokenSymbol})")
+                    append("and fee payer token ${feePayerToken.tokenSymbol}")
+                }
+            }
+
+            FeeCalculationState.Success(
+                fee = FeeRelayerFee(
+                    feesInSol = feesInSol,
+                    feesInFeePayerToken = feesInFeePayerToken,
+                    feesInSourceToken = feesInSourceToken,
+                    expectedFee = feesInSol
+                )
+            )
+        } catch (e: CancellationException) {
+            Timber.i("Fee calculation cancelled")
+            return FeeCalculationState.Cancelled
```

Comment:
multiple returns instead of one root return

Reply:
early return allows to avoid evaluating all the following conditions, but ok, I'll move zero check into `when`